Move UnconditionalSuppressMessageAttributes from xml to .cs files for Syndication Library

jtschuster opened the pull request:
Attempts to move the `UnconditionalSuppressMessageAttribute`'s from Suppressions.xml to the source files.

jtschuster committed:
Move suppressions to .cs files for Syndication Library

jtschuster committed:
Remove suppressions.xml

jtschuster committed:
Add definition for UnconditionalSuppressMessage for netstandard2.0

agocke commented on src/libraries/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/UnconditinalSuppressMessageAttribute.cs:
```diff
@@ -0,0 +1,93 @@
+#if NETSTANDARD2_0
+#nullable enable
+
+namespace System.Diagnostics.CodeAnalysis
```
Rather than adding another copy of this file, could you use a `<Compile Include="..."` to include the existing file from the corelib project?

MichalStrehovsky commented:
It doesn't appear this XML gets embedded in the shipping assembly. I think this one is one of those where we just suppress when IL Linker runs as part of the library build (in the non-shipping mode of IL Linker) and we don't want the warning to break the build.

Using an actual attribute would not be desirable here because we want this to still warn for the end user.

This is one of the places where we would need to duplicate what's in the XML into `#pragma warning disable` so that it shuts up the Roslyn analyzer (that won't look at the XML). Either that, or disable the trimming analyzer on this assembly. The IL Linking stage is going to still analyze this assembly so we wouldn't be losing much.

elinor-fung commented:
> It doesn't appear this XML gets embedded in the shipping assembly.

Do we actually have any ones that do this for the suppressions files? I only see descriptors, substitutions, and link attributes embedded by our targets:
https://github.com/dotnet/runtime/blob/1ba0394d71a4ea6bee7f6b28a22d666b7b56f913/eng/illink.targets#L97

The [repo doc for ILLink.Suppressions.xml](https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/ILLink-files.md#illinksuppressionsxml) makes it sound like the intent of all ILLink.Suppressions.xml files is to only suppress it in the runtime repo build, but still warn the end user.

jtschuster commented:
If we want to warn the end user but not warn in the library build, why do we not annotate with RUC except for the netstandard2.0 build? Or does the API need to have the same attributes for every target?

jtschuster committed:
Replace suppression attribute with #pragma

jtschuster committed:
restore the warnings after the line that raises IL2026

jtschuster commented:
I removed one `UnconditionalSuppressMessage` and added `#pragma warning disable IL2026` in the places where the build raises IL2026, but building the libraries with the build.cmd script still raises the warnings. Am I doing something wrong here?

elinor-fung commented:
Hm, our [public doc](https://docs.microsoft.com/dotnet/core/deploying/trimming/fixing-warnings#requiresunreferencedcode) says:

> `SuppressMessage` and `#pragma` directives are only present in source so they can't be used to silence warnings from the trimmer.

sbomer commented:
> The [repo doc for ILLink.Suppressions.xml](https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/ILLink-files.md#illinksuppressionsxml) makes it sound like the intent of all ILLink.Suppressions.xml files is to only suppress it in the runtime repo build, but still warn the end user.

I think you are right - the `ILLink.Suppressions.xml` files are just a baseline of existing warnings to make sure we don't introduce new ones, and the `ILLink.LibraryBuild.xml` ones are ones that we have intentionally decided to keep as warnings in the product.

sbomer commented:
> I removed one UnconditionalSuppressMessage and added #pragma warning disable IL2026 in the places where the build raises IL2026, but building the libraries with the build.cmd script still raises the warnings.

The pragma should silence warnings from the trim analyzer, but the linker will still need the suppression XML since the suppressions aren't permanent.
